lib/HLSL/LiveValueAnalysis.cpp

```diff
@@ -0,0 +1,1323 @@
+
+  std::replace(fullPath.begin(), fullPath.end(), '\\', '/');
+
+  return fullPath;
+}
+
+void LiveValueAnalysis::analyzeCFG() {
+  
+  // Find TraceRay call sites.
+  for( Function &main : *m_module ) {
+    for( Function &F : *m_module ) {
+
+      if( !F.getName().startswith( "dx.op.traceRay" ) )
+        continue;
+
+      for( User *user : F.users() ) {
+        CallInst *call = dyn_cast<CallInst>(user);
+        if( call && call->getParent()->getParent() == &main )
+          m_callSites.insert( call );
+      }
+    }
+  }
+
+  // Determine live values across each call.
+  std::set<BasicBlock *> phiSelectorSpills;
+
+  // Prepare Debug Info.
+  DebugInfoFinder DIF;
+  DIF.processModule(*m_module);
+
+  for (CallInst *callSite : m_callSites) {
+    // Determine CFG edges that can lead to this call site.
+    std::set<std::pair<Instruction *, Instruction *>> predCfgEdges;
+    {
+      BasicBlock *BB = callSite->getParent();
+      std::vector<BasicBlock *> worklist;
+
+      for (;;) {
+        for (auto i = llvm::pred_begin(BB); i != pred_end(BB); ++i) {
+          BasicBlock *pred = *i;
+          std::pair<Instruction *, Instruction *> edge =
+            std::make_pair(pred->getTerminator(), BB->begin());
+
+          if (!predCfgEdges.count(edge)) {
+            predCfgEdges.insert(edge);
+            worklist.push_back(pred);
+          }
+        }
+
+        if (worklist.empty())
+          break;
+
+        BB = worklist.back();
+        worklist.pop_back();
+      }
+    }
+
+    // Determine known branch conditions based on CFG edges.
+    // Prunes pred CFG edges based on invariants iteratively.
+    std::set<Value *> conditionsKnownTrue;
+    std::set<Value *> conditionsKnownFalse;
+
+    for (;;) {
+      bool change = false;
+      for (std::pair<Instruction *, Instruction *> edge :
+        std::set<std::pair<Instruction *, Instruction *>>(predCfgEdges)) {
+        BranchInst *branch = dyn_cast<BranchInst>(edge.first);
+
+        if (!branch || !branch->isConditional())
+          continue;
+
+        assert(branch->getNumSuccessors() == 2);
+
+        Instruction *taken = branch->getSuccessor(0)->begin();
+        Instruction *notTaken = branch->getSuccessor(1)->begin();
+
+        std::pair<Instruction *, Instruction *> takenEdge =
+          std::make_pair(edge.first, taken);
+        std::pair<Instruction *, Instruction *> notTakenEdge =
+          std::make_pair(edge.first, notTaken);
+
+        int count =
+          predCfgEdges.count(takenEdge) + predCfgEdges.count(notTakenEdge);
+
+        if (!count)
+          continue;
+
+        if (count == 2) {
+          if (conditionsKnownTrue.count(branch->getCondition())) {
+            predCfgEdges.erase(notTakenEdge);
+            change = true;
+          }
+          if (conditionsKnownFalse.count(branch->getCondition())) {
+            predCfgEdges.erase(takenEdge);
+            change = true;
+          }
+          continue;
```

Comment:
Removed continue to allow fall through.

Reply:
I'm not 100% sure I fully understand the code yet, but I don't think leaving the "continue" out is correct. If count == 2, the next conditional predCfgEdges.count(takenEdge) would always be true, which would result in the condition being treated as "always true", which in turn would result in the notTakenEdge being erased in a subsequent iteration.
I think the only reason why this could accidentally still work is because the pruning we do is incomplete (see other comment) and the only case where the removal would have an effect is if the edge was critical and there was a phi in the target block - and quite possibly we don't have any critical edges in the CFG at this point.

Reply:
Sounds reasonable to me; put continue back in for now.